[쇼핑 장바구니 3, 4단계] 이든 미션 제출합니다.

devfeijoa opened the pull request:
# 셀프 체크리스트

안녕하세요, 둘리 ! 이든입니다. 
3,4단계 리뷰도 잘 부탁드립니다.
늘 귀중한 시간내어 리뷰해주셔서 감사합니다 !

 <!--
 모든 항목을 점검한 후 리뷰를 요청하면, 보다 의미 있는 피드백을 받을 수 있습니다.
 -->

- [x] 프로그램이 정상적으로 작동하는가?
- [ ] 모든 테스트가 통과하는가?
- [ ] 이전에 받은 피드백을 모두 반영하였는가?
- [x] 코딩 스타일 가이드를 준수하였는가?
    - [x] IDE 코드 자동 정렬을 적용하였는가?
    - [x] 린트 검사를 통과하였는가?
- [x] 프로그래밍 요구 사항을 준수하였는가?
- [x] `README.md`에 기능 목록을 정리하고 명확히 기술하였는가?

# 어떤 부분에 집중하여 리뷰해야 할까요?

이번 미션에서 4단계는 거의 구현하지 못했습니다 😢 
3단계만 어찌저찌 구현한 상태여서 3단계 위주로 리뷰 부탁드려도 될까요?
최대한 빠르게 4단계도 구현하여 다음 요청에 반영하겠습니다 😭 

 <!--
 학습 목표에 맞게 구현하였는지 점검하고, 피드백을 받을 준비가 되었는지 확인하세요.
 정답을 묻기보다, 자신의 생각과 의사 결정 과정을 공유하며 피드백을 통해 논의하는 것이 좋습니다.
 -->
 
-

## 코드 리뷰 커뮤니케이션

 <!--
 리뷰어에게 질문이 있다면, PR에 직접 코멘트를 남겨 보세요.  
 특정 코드 줄에 의견을 남기면 보다 구체적인 피드백을 받을 수 있습니다.  
 -->

📌 [GitHub에서 PR에 댓글 남기는 방법](https://docs.github.com/en/pull-requests/collaborating-with-pull-requests/reviewing-changes-in-pull-requests/commenting-on-a-pull-request)

## 참고 자료

 <!-- 관련 문서나 참고할 만한 자료를 추가하세요. -->

## 스크린샷

 <!-- 필요하다면 스크린샷을 첨부하세요. -->

![]()

devfeijoa committed:
feat: 장바구니에 상품 추가 기능 구현

- 상품 목록 화면에서 '+' 버튼 클릭 시 상품 추가
- '+' 버튼 클릭 시 수량 조절 UI 표시
- 수량 조절 UI 디자인 및 기능 구현
- 관련 아이콘 추가

devfeijoa committed:
docs: step3 기능 명세서 업데이트

devfeijoa committed:
feat: id 추가

devfeijoa committed:
feat: ShoppingCartItem 구현

devfeijoa committed:
refactor: 주석 제거

devfeijoa committed:
feat: CartRepository 구현

devfeijoa committed:
refactor: id추가

devfeijoa committed:
refactor: id로 변경

devfeijoa committed:
refactor: CartRepository 개선

devfeijoa committed:
refactor: 배경색 추가

devfeijoa committed:
refactor: 상품 제거 클릭 리스너 인자 변경

devfeijoa committed:
feat: 장바구니 상태 모델 구현

devfeijoa committed:
feat: CartViewModel 수량 업데이트 기능 추가

devfeijoa committed:
feat: 상품 목록 화면에서 장바구니 추가/수량 조절 기능 구현

devfeijoa committed:
feat: 상품 목록에 장바구니 수량 반영

- 상품 상세에서 장바구니 담기 또는 삭제 시, 상품 목록 화면에도 해당 변경 사항이 반영되도록 개선

devfeijoa committed:
refactor: 사용하지 않는 메서드 제거

devfeijoa committed:
feat: total price 구현

devfeijoa committed:
feat: CartRepository에 상품 수량 조회 및 추가 기능 추가

devfeijoa committed:
feat: 상품 상세 화면 수량 조절 기능 추가

- 상품 상세 화면에서 수량을 조절하여 장바구니에 담을 수 있도록 UI 및 로직 추가
- `ProductDetailViewModel`에 수량 관련 LiveData 및 로직 추가
- `item_quantity.xml` 레이아웃을 사용하여 수량 조절 UI 구현
- 장바구니 아이템(`ShoppingCartItem`) 모델 변경: `Product` 대신 `ShoppingCartItem` 사용

devfeijoa committed:
feat: TextView에 수량을 표시하는 BindingAdapter 추가

devfeijoa committed:
feat: 수량 관련 xml 구현

devfeijoa committed:
refactor: 수량이 안 보이는 문제 개선

devfeijoa committed:
docs: 기능 명세서 업데이트

devfeijoa committed:
docs: step4 기능 명세서 작성

devfeijoa committed:
build: Room 의존성 추가

devfeijoa committed:
refactor: productId 타입 변경

- String -> Int

devfeijoa committed:
feat: ShoppingCartItemEntity 구현

devfeijoa committed:
feat: CartDao 구현

hyemdooly commented on app/src/main/java/woowacourse/shopping/data/ShoppingCartItemEntity.kt:
```diff
@@ -0,0 +1,41 @@
+                productPrice = shoppingCartItem.product.price,
+                productQuantity = shoppingCartItem.quantity,
+            )
+    }
```
확장함수로 빼기 vs 객체 내부에서 가지기
이든의 생각은 어떠신가요? 객체 내부에서 가지게된 이유가 있을까요?

hyemdooly commented on app/src/main/java/woowacourse/shopping/repository/CartRepository.kt:
```diff
@@ -0,0 +1,24 @@
+interface CartRepository {
+    fun addProduct(product: Product): CartState
+
+    fun addProduct2(cart: ShoppingCartItem)
```
정말 급하셨군요 ㅋㅋㅋ 😭

hyemdooly commented on app/src/main/java/woowacourse/shopping/model/products/CartState.kt:
```diff
@@ -0,0 +1,9 @@
+package woowacourse.shopping.model.products
+
+data class CartState(
```
상태라는 이름이 자연스러울까요? 🤔 
상태라기 보다는 장바구니 목록을 관리하는 친구로 보이네요

hyemdooly commented on app/src/main/java/woowacourse/shopping/repository/CartRepositoryImpl.kt:
```diff
@@ -0,0 +1,77 @@
+
+    companion object {
+        @Volatile
+        private var INSTANCE: CartRepositoryImpl? = null
```
변수명을 대문자로 지은 이유가 있으신가요?
상수인 경우에 대문자로 많이 짓습니다. 🤔

hyemdooly commented on app/src/main/java/woowacourse/shopping/view/cart/CartActivity.kt:
```diff
@@ -17,77 +17,90 @@ import woowacourse.shopping.viewmodel.cart.CartViewModel
-                            R.color.aqua_green,
-                        )
-                    ),
-                )
-            }
+            updatePageButtons()
+        }
```
여러 크루들이 pageCount를 관찰하고 있다가 버튼의 상태를 바꿔주도록 로직을 작성하고 있는데요.
버튼을 눌렀을 때, 카트 아이템을 삭제했을 때 두 경우에 버튼의 상태가 바뀐다고 정의할 수 있을 것 같아요.

그렇다면 pageCount가 바뀔 때 로직을 실행하는 것이 아니라, 위 두 가지 경우에 버튼의 상태를 바꿔주고
그 상태를 관찰하고 있다가 버튼 속성(배경색과 같은..)을 바꿔주는 것은 어떨까요?

hyemdooly commented on app/src/main/java/woowacourse/shopping/view/cart/CartAdapter.kt:
```diff
@@ -4,25 +4,21 @@ import android.view.LayoutInflater
 class CartAdapter(
-    private val itemsInCart: MutableList<Product> = mutableListOf(),
-    private val onProductRemoveClickListener: OnProductRemoveClickListener,
+    private val onProductRemove: (Int) -> Unit,
+    private val onQuantityChange: (Int, Int) -> Unit,
```
interface로 정의해볼 수 있을까요?

hyemdooly commented on app/src/main/java/woowacourse/shopping/view/cart/CartViewHolder.kt:
```diff
@@ -2,14 +2,30 @@ package woowacourse.shopping.view.cart
+    fun bind(item: ShoppingCartItem) {
+        binding.model = item
+
+        binding.onProductRemoveClickListener =
+            object : OnProductRemoveClickListener {
+                override fun onRemoveClick(productId: Int) {
+                    onProductRemove(productId)
+                }
+            }
+        binding.quantityControl.tvPlus.setOnClickListener {
+            onQuantityChange(item.product.id, item.quantity + 1)
+        }
+
+        binding.quantityControl.tvMinus.setOnClickListener {
+            if (item.quantity > 1) {
+                onQuantityChange(item.product.id, item.quantity - 1)
+            }
+        }
```
bind에서 세팅하지 않고, 초기화할 때 리스너를 variable로 넘기고
xml에서 사용할 수도 있습니다!
이전에는 그렇게 작성하셨던 것 같은데요 🤔

hyemdooly commented on app/src/main/java/woowacourse/shopping/view/products/ProductViewHolder.kt:
```diff
@@ -11,10 +12,40 @@ class ProductViewHolder(
     init {
         binding.productClickListener = productClickListener
+
+        binding.btnFloatingPlus.setOnClickListener {
+            val product = binding.product ?: return@setOnClickListener
+            productClickListener.onAddClick(product, 1)
+        }
+
+        binding.quantityControl.tvMinus.setOnClickListener {
+            val product = binding.product ?: return@setOnClickListener
+            productClickListener.onAddClick(product, -1)
+        }
+
+        binding.quantityControl.tvPlus.setOnClickListener {
+            val product = binding.product ?: return@setOnClickListener
+            productClickListener.onAddClick(product, 1)
+        }
```
여기도 마찬가지입니다!

hyemdooly commented on app/src/main/java/woowacourse/shopping/view/products/ProductsActivity.kt:
```diff
@@ -13,51 +14,89 @@ import androidx.recyclerview.widget.GridLayoutManager
-    private val viewModel: ProductsViewModel by viewModels { ProductsViewModel.Factory }
+
+    private val productsViewModel: ProductsViewModel by viewModels { ProductsViewModel.Factory }
+    private val cartViewModel: CartViewModel by viewModels { CartViewModel.Factory }
```
ProductsActivity에서 CartViewModel을 꼭 사용해야할까요? 🤔

hyemdooly commented on app/src/main/java/woowacourse/shopping/view/products/ProductsActivity.kt:
```diff
@@ -13,51 +14,89 @@ import androidx.recyclerview.widget.GridLayoutManager
                         startActivity(intent)
                     }
+
+                    override fun onAddClick(
+                        product: Product,
+                        quantity: Int,
+                    ) {
+                        if (quantity > 0) {
+                            cartViewModel.addToCart(product)
+                        } else {
+                            val currentQuantity = cartViewModel.getQuantity(product.id)
+                            cartViewModel.updateQuantity(product.id, currentQuantity - 1)
+                        }
+                    }
```
여기도 참고해주세요! 저도 의문이 들었어요.
수량이 0이 더라도 addToCart로 넣을 수 있지 않을까요?

hyemdooly commented on app/src/main/java/woowacourse/shopping/viewmodel/cart/CartViewModel.kt:
```diff
@@ -4,70 +4,111 @@ import androidx.lifecycle.LiveData
+
+    fun getQuantity(productId: Int): Int = _cartState.value?.getQuantity(productId) ?: 0
+
+    fun loadPage(page: Int) {
+        val cartState = _cartState.value ?: CartState()
+        val allItems = cartState.getAllShoppingCartItem()
+
+        if (allItems.isEmpty()) {
+            _currentPageItems.value = emptyList()
+            _pageCount.value = 1
+            return
+        }
+
+        val totalPages = (allItems.size + pageSize - 1) / pageSize
+        val validPage = page.coerceIn(1, totalPages)
+
+        val start = (validPage - 1) * pageSize
+        val end = minOf(start + pageSize, allItems.size)
+
+        val pageItems = allItems.subList(start, end)
+        _currentPageItems.value = pageItems
+        _pageCount.value = validPage
```
전체 아이템을 가져와서 subList하는 것은 pagination일까요? 🤔

hyemdooly commented on app/src/main/java/woowacourse/shopping/viewmodel/cart/CartViewModel.kt:
```diff
@@ -4,70 +4,111 @@ import androidx.lifecycle.LiveData
+    fun isFirstPage(): Boolean = (_pageCount.value ?: 1) == 1
 
-    fun isLastPage(pageCount: Int): Boolean {
-        val totalPageCount = (cart.productsInCart.size + pageSize - 1) / pageSize
-        return pageCount == totalPageCount
+    fun isLastPage(): Boolean {
+        val cartState = _cartState.value ?: CartState()
+        val allItems = cartState.getAllShoppingCartItem()
+        val totalPages = if (allItems.isEmpty()) 1 else (allItems.size + pageSize - 1) / pageSize
+        return (_pageCount.value ?: 1) == totalPages
```
버튼의 상태를 판단하기 위해 가지고 있는 함수라면, LiveData로 상태를 만들고 버튼이 그 변수를 바라보도록 하면 어떨까요?

hyemdooly reviewed (changes requested):
이든 안녕하세요! 일이 많아서 이제서야 리뷰를 드립니다. 너무 늦었네요..🙏
4단계 양이 더 많을 것 같은데 시간 괜찮으신가요? 😭 
고민해볼 부분들, 궁금한 부분들에 리뷰를 드렸는데요! 어떻게 하면 뷰와 LiveData를 잘 연결해서 활용해볼 수 있을지 고민해보면서 리뷰를 반영하시면 좋을 것 같아요.
고생 많으셨어요!! 다음 요청 때는 4단계까지 완성한 코드를 보고 싶네요.ㅎㅎ 화이팅입니다 💪

devfeijoa committed:
feat: CartDao 구현

devfeijoa committed:
feat: ShoppingCartEntity 구현

devfeijoa committed:
feat: ShoppingCartDatabase 구현

devfeijoa committed:
refactor: int로 타입 변경

devfeijoa committed:
rename: shoppingCart로 네이밍 변경

devfeijoa committed:
feat: ShoppingCartDataSource 구현

devfeijoa committed:
refactor: ROOM 데이터베이스 장바구니 구현 및 DataSource 구현

devfeijoa committed:
refactor: 장바구니 페이지네이션 기능 개선

devfeijoa committed:
refactor: `addCart` 메서드명을 `updateCart`로 변경

devfeijoa committed:
feat: Application 클래스 구현 및 AndroidManifest에 등록

devfeijoa committed:
feat: ProductDiffUtil 구현

devfeijoa committed:
refactor: CartViewModelFactory 구현

devfeijoa committed:
refactor: ProductsAdapter를 ListAdapter로 변경 및 관련 로직 개선

devfeijoa committed:
refactor: ProductDetailViewModel 팩토리 구현 및 UI 로직 개선

- ProductDetailViewModelFactory를 구현하여 ViewModel 생성 방식을 개선했습니다.
- ProductListViewType에 quantity 필드를 추가하여 상품 수량을 관리하도록 변경했습니다.
- ProductViewHolder에서 상품 수량 UI 업데이트 로직을 개선했습니다.
- ProductsAdapter에서 ProductViewHolder 바인딩 로직을 수정했습니다.
- ProductDetailActivity에서 ViewModelFactory를 사용하여 ProductDetailViewModel을 주입받도록 변경했습니다.
- item_product.xml 레이아웃에서 데이터 바인딩 변수명을 product에서 item으로 변경하고, quantity 바인딩을 추가했습니다.
- ProductsActivity에서 ViewModel observe 로직 및 수량 새로고침 로직을 일부 주석 처리했습니다. (추후 개선 예정)

hyemdooly reviewed (approved):
이든 안녕하세요~! 쇼핑 장바구니 3, 4단계 미션도 고생 많으셨어요.
안드로이드가 처음이라 미션 구현하기 급급했을텐데, 그 과정 속에서도 얻어가는 것이 있으셨기를 바라요!
하지만 실무에서는 기한 내에 구현하지 못하는 것은 있을 수 없는 일입니다. 😢 레벨 3, 4에는 정해진 기간 안에 구현하는 연습을 꼭 해보시길 바랍니다!! 💪 

앞으로의 우테코 여정도 응원하겠습니다!! 😄
